### ISSUE TITLE
Multi-file upload only uploads 1 file to the order due a file dialog issue

### ISSUE THREAD

**Issue by davidsherlock** [easydigitaldownloads/edd-upload-file#27]
To replicate:

1. Mark a product as requiring files. Make sure you set the file limit to more than one.

2.  Add the product to the cart.

3. Select the upload button at the checkout and multi-select 3 files within the file dialog box. The 3 files appear to be uploaded without problems (I used images). The thumbnails show, etc.

4. Complete the order.

5. Only 1 file has been uploaded.

**Comment by davidsherlock** [comment]
In fact, uploading multiple files dosen't work at all for some reason - regardless of where (checkout or receipt), only the first file is ever uploaded.

**Comment by SDavisMedia** [comment]
Replicated. I have a new ticket about it.

Likewise, if you go to Downloads -> Settings -> Extensions -> Upload File, and set the max number of uploads to 0, which would mean there is no max, the value appears to save as 1 (someone replicate that please).

I think we need to do a check up on the functionality of this extension.

**Comment by pippinsplugins** [comment]
@sunnyratilal could you please tackle this after CFM conditionals?

**Comment by sunnyratilal** [comment]
Yep, I'll take a look

**Comment by sunnyratilal** [comment]
@SDavisMedia I'll create a new issue for the bug you mentioned about allowing unlimited files.

**Comment by sunnyratilal** [comment]
Looks like it's a bug with the way the meta is being saved:
![image](https://user-images.githubusercontent.com/1525554/31665352-7675811a-b340-11e7-8e00-e75b3495433d.png)

**Comment by sunnyratilal** [comment]
Ok so all the files are uploaded locally on to the server, the meta schema is broken meaning we have multiple `edd_upload_file` meta keys uploaded for a single post ID.

We're going to need an upgrade routine to fix all this broken meta.

**Comment by sunnyratilal** [comment]
Upgrade routine needs to do the following:

- Remove all duplicate meta
- Concatenate all the file upload data together into a single row

Then we need to rewrite the code that displays the file uploads to work with the changed meta schema.

**Comment by sunnyratilal** [comment]
Upgrade routine successfully changes the meta schema 🎉 
![image](https://user-images.githubusercontent.com/1525554/31677130-bf2debd8-b361-11e7-872f-ba31a1095200.png)

**Comment by sunnyratilal** [comment]
PR opened at #29

**Comment by cklosowski** [comment]
Testing now @sunnyratilal

**Comment by cklosowski** [comment]
I've never used Upload file on my test site...is it possible to not show this upgrade routine if it's not necessary @sunnyratilal? Maybe just check to see if there are any rows with the `meta_key`, and if not mark the upgrade as complete?

![screen shot 2017-10-17 at 2 51 51 pm](https://user-images.githubusercontent.com/1390124/31691366-bb188fc4-b34a-11e7-8a61-d75e1006c0bb.png)

**Comment by sunnyratilal** [comment]
@cklosowski Good point, added in https://github.com/easydigitaldownloads/edd-upload-file/commit/cb9aec8e2c9dc6c90a8019ef1fed5e1c87958f65

**Comment by cklosowski** [comment]
@sunnyratilal on Nginx, I'm getting this error when trying to complete a purchase with an upload
```
{
error: "Server error. Not a multipart request. Please set forceMultipart to default value (true).",
uploadName: null
}
```

It's happening for me on master too...wonder if it's my config.

**Comment by cklosowski** [comment]
As a note, it also happens with Apache as my server.

**Comment by cklosowski** [comment]
Ignore me, this was an issue conflicting with Checkout Fields Manager. When that has errors it seems that the request completely fails. I'll open a new ticket for that.

**Comment by cklosowski** [comment]
Ok single files uploaded fine.

I tried 3 files (for a 3 file max product) and upon purchase completion, I saw these two errors (twice each)
`Invalid argument supplied for foreach() | wp-content/plugins/edd-upload-file/includes/actions.php:115`

`Illegal string offset 'filename' | wp-content/plugins/edd-upload-file/includes/actions.php:125`

And my purchase confirmation looks like this:
![screen shot 2017-10-18 at 9 30 48 am](https://user-images.githubusercontent.com/1390124/31730539-0f9570ac-b3e7-11e7-8f28-401b753a3fb5.png)

These were 2 .zip files and one .png.

**Comment by cklosowski** [comment]
Checking the DB @sunnyratilal and for this payment `3707` I have the following entries for the upload file data:
![image 2017-10-18 at 9 32 18 am](https://user-images.githubusercontent.com/1390124/31730652-6e40affe-b3e7-11e7-9799-3c791b596b5a.png)

**Comment by sunnyratilal** [comment]
@cklosowski ready for retesting

**Comment by cklosowski** [comment]
Seems to be working great for me. @easydigitaldownloads/core-devs would like 1 more test to verify and then I'll merge it in.

**Comment by pippinsplugins** [comment]
Should this be adding an upload field for each quantity in the cart?
<img width="728" alt="screen shot 2017-10-24 at 9 19 48 am" src="https://user-images.githubusercontent.com/1034109/31948428-8af06a5e-b89c-11e7-9cd6-c0bb8e348506.png">

Something didn't work here for me. I uploaded two files to one of the products and one file to another. On the receipt I see nothing. On view order details the metabox is empty:
<img width="983" alt="screen shot 2017-10-24 at 9 21 17 am" src="https://user-images.githubusercontent.com/1034109/31948541-d03be1a6-b89c-11e7-89dd-d3fe86e721f6.png">

**Comment by sunnyratilal** [comment]
I talked to @cklosowski about this and albeit it seems _odd_ that are there are separate upload fields for each quantity, it does make sense.  If we take the example of a store providing a logo design service, should a customer want 10 logos then they would upload files for each logo.

I'm all ears to making the UI more intuitive but I think that breaches the scope of this specific issue where we are currently just trying to ensure multiple file uploads work?

The meta schema does however need slightly tweaking again to allow for quantity support; I'll get that fixed.

**Comment by mintplugins** [comment]
I think to me it makes sense to have an upload field for each quantity.

**Comment by pippinsplugins** [comment]
:+1:

**Comment by cklosowski** [comment]
@sunnyratilal let me know when this is ready and I'll test it out for you.

**Comment by pippinsplugins** [comment]
@sunnyratilal could you please take another stab at this one so we can finish out the release?

**Comment by sunnyratilal** [comment]
@pippinsplugins of course, will be online in an hour or so and will prioritise.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by sunnyratilal** [comment]
@pippinsplugins could you test again please?

I'm now able to upload files to an item with quantity set to 10 and then view them on the purchase receipt and on the admin side.

**Comment by pippinsplugins** [comment]
@sunnyratilal works great now!

We have a merge conflict though, could you resolve that please?

**Comment by pippinsplugins** [comment]
Disregard that, it's not working for me. I tested on `master` by accident.

On `issue/27` I get JS errors:
![image](https://user-images.githubusercontent.com/1034109/33741919-df8c5b3a-db6b-11e7-9f3c-2c9313a4aa2b.png)
